tfx/components/example_gen/utils.py

```diff
@@ -209,84 +213,183 @@ def _glob_to_regex(glob_pattern: Text) -> Text:
-                       (file_path, split_regex_pattern))
+      raise ValueError('Cannot find %s number from %s based on %s' %
+                       (SPAN_PROPERTY_NAME, file_path, split_regex_pattern))
+
+    select_version = None
+    if is_match_version:
+      select_version = result.group(VERSION_PROPERTY_NAME)
+      try:
+        version = int(select_version)
+      except ValueError:
+        raise ValueError('Cannot find %s number from %s based on %s' %
+                         (VERSION_PROPERTY_NAME, file_path,
+                          split_regex_pattern))
+
     if latest_span is None or span >= int(latest_span):
       # Uses str instead of int because of zero padding digits.
-      latest_span = result.group(1)
+      latest_span = select_span
+      latest_version = select_version
+    if latest_version is None or version >= int(latest_version):
+      latest_version = select_version
 
-  if latest_span is None:
-    raise ValueError('Cannot not find matching for split %s based on %s' %
+  if latest_span is None or (is_match_version and latest_version is None):
+    raise ValueError('Cannot find matching for split %s based on %s' %
                      (split.name, split.pattern))
-  return latest_span
+
+  return latest_span, latest_version
+
+
+def _retrieve_latest_span_version(uri: Text,
+                                  split: example_gen_pb2.Input.Split
+                                  ) -> Tuple[Optional[Text], Optional[Text]]:
+  """Retrieves the most recent span and version for a given split pattern.
+
+  If both Span and Version spec occur in the split pattern, searches for and
+  returns both the latest Span and Version. If only Span exists in the split
+  pattern, searches for the latest Span, and Version is returned as None.
+  If Version is present, but not Span, an error is raised. If neither Span
+  nor Version is present, returns both as None.
+
+  Args:
+    uri: The base path from which files will be searched.
+    split: An example_gen_pb2.Input.Split object which contains a split pattern,
+      to be searched on. Note that this function will update the {SPAN} in this
+      and {VERSION} tags in the split config to actual Span and Version numbers.
+  Returns:
+    Tuple of two strings, Span (optional) and Version (optional).
+
+  Raises:
+    ValueError: if any of the following occurs:
+      - If either Span or Version spec is occurs in the split pattern
+        more than once.
+      - If Version spec is provided, but Span spec is not present.
+      - If Span or Version found is not an integer.
+      - If a matching cannot be found for split pattern provided.
+  """
+
+  latest_span = None
+  latest_version = None
+
+  if SPAN_SPEC in split.pattern:
+    if split.pattern.count(SPAN_SPEC) != 1:
+      raise ValueError('Only one %s is allowed in %s' % (SPAN_SPEC,
+                                                         split.pattern))
+
+    is_match_version = VERSION_SPEC in split.pattern
```

Comment:
Will there be any problem if we always optimistically try to get the version info? We don't need the `is_match_version` then.

Reply:
I think that it would just move this `is_match_version` boolean inside the `_glob_and_match_span_version()`. Since we already do all the processing to check if the `{VERSION}` spec is present in  `_retrieve_latest_span_version()`, I think that would repeat code if we were to optimistically try to get version info in the  `_glob_and_match_span_version()`.

On a structure note, I think having `_retrieve_latest_span_version()` as verifying the specs in the pattern and deciding what to match in the glob (like it does right now through the boolean) makes more sense than moving the spec verification inside  `_glob_and_match_span_version()`. @hughmiao any opinion?

Reply:
I feel we can merge this function and _glob_and_match_span_version

if we move line 332~336 to calculate_splits_fingerprint_span_and_version (check the comments below)

this function is basically just check existance and count, and it's pretty much align with the code structure in _glob_and_match_span_version line 223~234, consider merge them?

Reply:
latest draft of PR has them merged now